Platform: Add C++ Span class.

pan- opened the pull request:
### Description

The Span class allows the creation of views over contiguous memory. The view
do not own memory, is typed and has a length. It can be used as a replacement of
the traditional pair of pointer and size in parameters or class fields.

Main operations:
- size(): return the lenght of the memory viewed
- empty(): return if the memory viewed is empty
- [index]: access elements viewed
- data(): return a pointer to the memory viewed.
- first(count): Create a subview from the first count elements.
- last(count): Create a subview from the last count elements.
- == and !=: compare two views or a view to array and return if they are equal or
not.

The Span class came in two flavors:
- Static size: The size is encoded in the Span type and it is as lightweitgh as
a single pointer,
- Dynamic size: The object can store arbitrary views and it costs one pointer
and the size of the view.

### Pull request type


    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Breaking change

pan- committed:
Platform: Add C++ Span class.

The Span class allows the creation of views over contiguous memory. The view
do not own memory, is typed and has a length. It can be used as a replacement of
the traditional pair of pointer and size in parameters or class fields.

Main operations:
- size(): return the lenght of the memory viewed
- empty(): return if the memory viewed is empty
- [index]: access elements viewed
- data(): return a pointer to the memory viewed.
- first(count): Create a subview from the first count elements.
- last(count): Create a subview from the last count elements.
- == and !=: compare two views or a view to array and return if they are equal or
not.

The Span class came in two flavors:
- Static size: The size is encoded in the Span type and it is as lightweitgh as
a single pointer,
- Dynamic size: The object can store arbitrary views and it costs one pointer
and the size of the view.

pan- commented:
@donatieng @paul-szczepanek-arm This is an improved version of the `Arrayview` class we have use in BLE for now quite some time. It is also used by the NFC Ndef layer.

paul-szczepanek-arm commented:
why not update array view? do we really need both of them?

pan- commented:
`ArrayView` is a private API of ble that serves the same purpose. This one is a public API of mbed OS; the plan is to replace the BLE ArrayView by this one if it gets in. Changing the name to `Span` is more in line with C++ standard; and it is easier to type too. 

Two functions have been added: `first` and `last` that create subspans. New comparison exists too; it allows programmer to write statement such as: 

```c++ 
static const char hello[] = "Hello World!";

bool is_greeting(const Span<uint8_t>& v) { 
    return v == hello;
}
```

pan- committed:
Span: Cleanup usage of ptrdiff_t and size_t

pan- committed:
Span: Improve comparison operator to array by using fixed size Span.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+template<typename T, ptrdiff_t Size = SPAN_DYNAMIC_SIZE>
+struct Span {
+
+    MBED_STATIC_ASSERT(Size >= 0, "Invalid size for an ArrayView");
```
Search-and-replace fail on the name.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+     */
+    bool empty() const
+    {
+        return size() ? false : true;
```
Weird implicit boolean inside ternary operator. Surely `size() == 0`?

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+     *
+     * @return The raw pointer to the array.
+     */
+    const T* data() const
```
Formatting rules say pushing the `*&` to the right.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+     * @param array_size Number of elements of T present in the array.
+     *
+     * @post a call to size() will return array_size and data() will return
+     * array_tpr.
```
(`@p`)? `array_ptr`

kjbracey reviewed (commented):
Looks good to me, apart from editing nits. I've been advocating `string_view` for a while (mainly to client guys) - not sure if this effectively fully covers that, or it would need a few more implicit conversions.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+ * If the type use this value then the size of the array is stored in the object
+ * at runtime.
+ */
+#define SPAN_DYNAMIC_SIZE -1
```
Just for more naming consistency with C++, should the the template parameter and this constant be "extent" rather than "size"?

Feels a bit wonky, but I guess it's intended to allow you to separately query `extent` from `size` in template nonsense. (`if (span.extent == std::dynamic_extent) ...`)

pan- replied:
I will fix the name of the dynamic tag however it is not meant to be queried at runtime as this information is not really useful. 

Span with fixed extent are mostly useful for type(def) that requires a fixed size otherwise the dynamic extent should be used. Also, the conversion rules offer implicit conversion from `Span<T, Size>` to `Span<T>` so users don't have to worry about conversions.

pan- committed:
Span: Make name of dynamic span tag consistent with C++ standard.

pan- committed:
Span: Fix odd condition in Span::empty.

pan- committed:
Span: Fix typo.

pan- committed:
Span: Make type pointer and reference declaration consistent with guideline.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+ * SPAN_DYNAMIC_SIZE  is special as it allows construction of Span objects of
+ * any size (set at runtime).
+ */
+template<typename T, ptrdiff_t Size = SPAN_DYNAMIC_EXTENT>
```
In C++ the template parameter is also `Extent`.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+     * @param array_ptr Pointer to the array data
+     * @param array_size Number of elements of T present in the array.
+     *
+     * @post a call to size() will return array_size and data() will return
```
That's not true - you'll return `Size`, not `array_size`. 

C++ restricts this constructor to matching size, making the difference only visible in undefined behaviour. Tighten the assert?

pan- replied:
I'm not sure to understand why this is still up has it has been fixed in https://github.com/ARMmbed/mbed-os/pull/7828/commits/3985fb8d62adbe859c959f4e6f4444ed6958f4a4

pan- commented:
Thanks @kjbracey-arm ; I've updated my code according to your review. 

To have a `string_view` we should have a `string` first! More seriously, even if string functions are not present inside the class, the `Span` class can already do a lot in SAX type parsers where tokens parsed can be propagated to users without copy. 

I think it could be easy to add _find_ functions as free functions targeting Span<char>; all other aspects of `string_view` are covered by Span.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,578 @@
+     */
+    size_t size() const
+    {
+        return _array ? Size : 0;
```
Do you really need a static span to have the runtime option of being empty? C++ only permits empty for `Extent==0` or dynamic spans. From a compile-time correctness view, I quite like the idea of enforcing non-0 static spans to be non-null.

pan- replied:
I think it is valuable to not enforce non-0 static spans to be non-null as it can be use to represent the absence of values. That's useful for optional values represented by a static span. 

```c++
struct pdu_t { 
    // options have a fixed size 
    typedef Span<uint8_t, 5> option_t;

    uint8_t header;
    option_t option;
    Span<uint8_t> payload;
};

void process_pdu(const pdu_t& pdu) { 
    if (!pdu.option.empty()) { 
        // process options ...
    }
}
```

Note that it raises the question of `operator bool() const`. It is not part of `std::span` for consistency with the standard library (see [1](https://github.com/Microsoft/GSL/issues/202) and [2](https://github.com/Microsoft/GSL/issues/140)); I believe that is up to debate in our use case; but maybe we can have this debate once this class gets in and have more feedbacks from users.

kjbracey replied:
Ta for links. I can see it's useful, much as allowing references to be null would be useful, but it is a major deviation from the standard C++ equivalent, where that syntax guarantees there are 5 `uint8_t`s.

I believe the C++ standard guys would tell you to use `optional<span>`, but then they're not too bothered about doubling memory usage by adding an extra `bool` alongside a pointer. In the absence of `optional`, and to save space, maybe you could designate the extended type `OptionalSpan` to make clear the extended semantics, although it becomes less catchy.

pan- replied:
If we ever have an optional type (would you be interested by one ? ), we can have a template specialization for `Span` that only stores the `Span` and look at `data()` to determine if the array is present or not. 

The issue with consistency is that depending on the stand point something can be declared consistent or not given that the language is full of inconsistencies. 

An array naturally decays to a pointer in pointer/boolean context so from that perspective it would be consistent to have the `operator bool() const` member function. At the same time, standard/gsl members consider that `span` should be treated as a regular container and none of the container exposes that operator: they do expose an `empty()` member function to test if the container is empty or not. 

----

To go back, to the initial issue you raised, in `std::span`, it is perfectly fine to store nothing in fixed size span. Should the code in this PR change to not accept empty span of non-0 size ?

kjbracey replied:
Agree that the language is utterly full of inconsistencies, and it's alarming that we see stuff like `string_view`->`span` being thrashed out in consecutive standards that are (from our point of view) 3-4 standards in the future. `string_view` being semi-obsolete before we ever get it is just daft.

But whenever they _do_ manage to pin something down, if we're going to copy its name I think we should copy the full semantics, just to not add to the confusion.

I'm also with Herb Sutter on the importance of [distinguishing](https://herbsutter.com/2017/02/15/distinguishing-between-maybe-null-vs-never-null-is-the-important-thing/) "maybe NULL" and "not-NULL" in the language.

In part that's because I'm aware that if the language could enforce non-NULL for pointers we could get rid of hundreds of runtime checks for NULL that get inserted to cope with unit tests that try passing NULL (sometimes as their only "test"). Unit tests may not believe in argument preconditions, but if we can stop them compiling...

So when I see that the standard has left room for that distinction, I'd like to have the chance to use it. In this form, the PR leaves no nice namespace room for the alternative "non-NULL" semantics. I'd like to reserve the name `Span` for that, with the alternative being `OptionalSpan` or equivalent. Then whether this PR actually adds `Span` or `OptionalSpan` or both is up to you - whatever you think you currently need?

And `OptionalSpan` could certainly include `bool()` as `optional<span>` does.

kjbracey reviewed (commented):
Few more review notes.

kjbracey commented:
> To have a string_view we should have a string first! 

Indeed! The client guys do have an internal `String` they use, but they've been gradually stripping it out  in favour of pointer+length to save memory. I've previously suggested that a `StringView` would solve most of their problems.

It seems your implementation does have helpers like the array<->span comparison that C++ lacks - C++ only has those for `string_view`, afaict.

pan- committed:
Span: replace reference to Size with Extent.

pan- committed:
Span: Fix documentation error.

pan- commented:
Standard C++ defines standard comparison operator for spans however it does not define helpers that auto convert arrays into span: 

```c++
uint8_t foo[] = "hello";

// works
if (mbed::Span<uint8_t>(foo) == foo) { }

// error
if (std::span<uint8_t>(foo) == foo) { }

// works
if (std::span<uint8_t>(foo) == std::span<uint8_t>(foo)) { }
```

pan- committed:
Span: Improve consistency with standard.

This commit aims to make Span implementation more in line with what is present in N4762:
- use appropiate index types where applicable.
- use typedefed type inside the class (index_type, reference, pointer, element_type)
- assertion where applicable
- restrict default construction to Span with extent == 0 or extent == dynamic.
- construct span from a range of pointer
- remove non const overload of the subscript operator
- remove non const overload of the data function
- implement subspan function
- implement missing first and last function of dynamic span

pan- commented:
@kjbracey-arm following our offline discussion, I've updated the implementation. to be more in line with the latest standard draft. Apologies if I haven't split it into many commits :(.

pan- committed:
Span: Fix subspan return type

pan- committed:
Span: Improve readability of precondition checks

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,795 @@
+     * Extent != 0 .
+     */
+    Span() : _data(NULL) {
+        MBED_STATIC_ASSERT(Extent == 0, "Invalid extent for a Span");
```
Shame we can't kill the overload suppress. Anyway, this isn't the correct error - should be something like "Cannot default construct a static-extent Span (unless Extent is 0)". Make that less wordy if you can...

Mind you, suppressing the overload would lead to a less-clear error message.

pan- replied:
It is possible to kill the overload (see [here](https://github.com/Microsoft/GSL/blob/1995e86d1ad70519465374fb4876c6ef7c9f8c61/include/gsl/span#L379-L384)) but if I can stay away of enable_if and other TMP tricks I do as others may have to touch at the code later. 

All in all, static assert kills the overloads if it is selected for instantiation. Given that it is a nullary function and there is only one nullary overload per function name, the only case where the overload is selected is when the default constructor is called. In the end, that doesn't change much to kill it from the overloading set or to kill it if it is instantiated. 

Side note: clang implementation use the same construct.

kjbracey replied:
> others may have to touch at the code later.

Quite :)

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,795 @@
+     */
+    reference operator[](index_type index) const
+    {
+        return _data[index];
```
Would this be a good place for an `MBED_ASSERT` range-check on `index`? Though maybe that should be limited to debug builds, not normal develop.

pan- replied:
I thought about it, that is the place where it potentially costs the most. MBED_ASSERT are preserved in develop builds. 

Here is some data about the cost of checks: https://godbolt.org/z/E5c1b6 
MINIMAL_CHECK tests the upper boundary while FULL_CHECK tests  the lower and upper boundary.

kjbracey replied:
Yes, I think it's too much for develop builds. Could do it conditionalised on `MBED_DEBUG` or a custom flag. Anyway, can be added later.

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,795 @@
+     * @param count Number of elements viewed.
+     *
+     * @pre [ptr, ptr + count) must be be a valid range.
+     * @pre count must be equal to extent.
```
Extent

kjbracey commented on platform/Span.h:
```diff
@@ -0,0 +1,795 @@
+        _data(elements) { }
+
+    /**
+     * Return the size of the array viewed.
```
Quite a few references to "array" or "array viewed" in comments where it should now be span.

kjbracey reviewed (commented):
This looks a lot better to me. Just editing issues remain, I think.

pan- committed:
Span: Fix static assert message in default constructor.

pan- committed:
Span: add boundary check in subscript operator for debug profile

pan- committed:
Span: amend documentation

pan- commented:
@kjbracey-arm I've refreshed the documentation and fixed your other comments.

cmonr commented:
@pan- Would you mind taking a look at the Travis CI docs failure?

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ * The size of the sequence can be encoded in the type itself or in the value of
+ * the instance with the help of the template parameter Extent:
+ *
+ *   - Span<uint8_t, 6>: Span over a sequence of 6 element
```
elements.

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ *   - Span<uint8_t>: Span over an arbitrary long sequence.
+ *
+ * When the size is encoded in the type itself, it is guaranteed that the Span
+ * view a valid sequence (not empty() and not NULL). The type system also prevent
```
if Extent != 0 / unless Extent == 0

prevents

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ *
+ * When the size is encoded in the type itself, it is guaranteed that the Span
+ * view a valid sequence (not empty() and not NULL). The type system also prevent
+ * automatic conversion from Span of different sizes. Finally, the size of the
```
use of size possibly confusion here. Storage/representation or something?

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ * When the size is encoded in the type itself, it is guaranteed that the Span
+ * view a valid sequence (not empty() and not NULL). The type system also prevent
+ * automatic conversion from Span of different sizes. Finally, the size of the
+ * span object is a single pointer.
```
Span

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ *
+ * When the size of the sequence viewed is encoded in the Span value, span
+ * instances can view invalid sequence (empty and NULL pointer). The function
+ * empty() helps client code to decide if valid content is being viewed or not.
```
Wording a little odd - you're not viewing an invalid sequence, you're viewing an empty sequence, which is valid. And an empty sequence can have a NULL pointer, and that's still valid. I think.

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ * @paragraph Example
+ *
+ * - Encoding fixed size array: Array values in parameter decays automatically
+ * to pointer which leaves room for subtitle bugs:
```
subtle

kjbracey commented on platform/Span.h:
```diff
@@ -33,29 +33,140 @@ namespace mbed {
+ * void process_mac(mac_address_t);
+ *
+ * // compile just fine
+ * uint8_t* invalid_value = NULL;
```
Examples should probably have `*` on right like real code.

pan- committed:
Span: Fix doxygen tags.

pan- commented:
@cmonr Fixed.

pan- committed:
Span: Fix documentation.

pan- commented:
@kjbracey-arm I think I've addressed your comments on the documentation.

pan- committed:
Span: Use mbed way of writing types in documentation example.

pan- commented:
@kjbracey-arm I think I will never get used to that way of writing of types 😭 . It is updated now. 
removing `*` in code block makes code more readable too (and it can be copy pasted) .

0xc0170 commented:
> @kjbracey-arm I think I will never get used to that way of writing of types 😭 . It is updated now.
removing * in code block makes code more readable too (and it can be copy pasted) .

Can you review astyle job in  Travis (shows few lines there)

pan- committed:
Span: Fix opening brace position.

pan- commented:
I fixed opening braces but some suggestions are wrong and inconsistent as astyle - unlike clang format - does not have the syntactic context : 

Confused about reference to array: 

```diff 
-    Span(element_type (&elements)[Count]):
+    Span(element_type(&elements)[Count]):
```

Confuse arithmetic operation and template type: 

```diff 
-    Span<element_type, Count == SPAN_DYNAMIC_EXTENT ? SPAN_DYNAMIC_EXTENT : Count>
+    Span < element_type, Count == SPAN_DYNAMIC_EXTENT ? SPAN_DYNAMIC_EXTENT : Count >
```

I'm not sure to understand what it tries to do: 

```diff 
         return Span<element_type, SPAN_DYNAMIC_EXTENT>(
-            _data + (_size - count),
-            count
-        );
+                   _data + (_size - count),
+                   count
+               );
```

0xc0170 commented:
First two are known limitations (as you noted, does not know the syntactic context 😢 ), therefore I would turn off astyle for these lines (templated classes like this).

It's correct most of the time. if not, turn it off for cases like this.

pan- committed:
Span: Allow copy construction from convertible span.

Addition of these overloads help when Span<const T> is constructed from Span<T>.

pan- commented:
@kjbracey-arm I realized it wasn't possible to achieve the following construct: 

```c++
Span<T> buffer;
Span<const T> const_buffer(buffer);
```

It is implemented in https://github.com/ARMmbed/mbed-os/pull/7828/commits/d5051a8ca7acf5d232f0eea257e820fa5c1689c4 which relies on SFINAE. Using a traits to remove the constness in the signature is not viable option and static assertion may have undesirable side effect.

cmonr commented:
@ARMmbed/mbed-os-core Could someone prioritize reviewing this tomorrow/today? Just found out/realized that this blocks https://github.com/ARMmbed/mbed-os/pull/7822

pan- committed:
Span: use static assert to kill copy construction from an incompatible span type.

Copy construction between Span of compatible type is allowed to fulfil the use
case Span<T> -> Span<const T>. This is achieved by a templated copy constructor
like constructor.

In p0122, the overload is discarded from the constructor set if the ElementType
of the Span in input is not convertible into the ElementType of the Span being
constructed.

To discard function overload, SFINAE has to be used which polutes the documentation
and make the code harder to read and maintain.

Unlike p0122, our Span class doesn't exposes (yet) functions with default argument
or functions that convert container in input into span the only overload with the
a single parameter that we exposes are:
- template<size_t N> Span(ElementType (&element)[N])
- Span(const Span& other): <- generated by the compiler.

For both of this functions we expect exact match and their resolution should not
interfere with the constructor that converts from another type of Span.

As a result it is possible to rely solely on C++ default resolution rules as we
won't hit cases were constructors convert from another type (std::array, std
container, span) and raise an error with a static assert if the element type
can't be converted.

If another copy - conversion - constructor is added then SFINAE has to be
reintroduced.

pan- commented:
@kjbracey-arm I evaluated a bit more the static assert option and it should be good as long as we have a single conversion constructor. SFINAE would have to be reintroduced if we add conversion functions from other type of containers. https://github.com/ARMmbed/mbed-os/pull/7828/commits/af69e1fb8bfc2c4f53abc0bffed852a3394eb245 commit message contains a more detailed explanation.

0xc0170 commented:
@pan- I can't locate tests and documentation PR for this addition

pan- commented:
@0xc0170 The number of tests we can set are rather limited and trivial as the type is just a constrained pointer (and optionally a size) to a contiguous sequence. Precondition violation will either generate a failure are runtime or compile time and we cannot test these unfortunately. I can add some _test_ but they will have to be converted to the new unit test framework later.

I will open an entry in the handbook; the class documentation is a good start.

Commit:
Copy edit Span.h

Copy edit file, mostly for consistent capitalization, punctuation and tense.

AnotherButler commented on platform/Span.h:
```diff
@@ -0,0 +1,983 @@
+/**
+ * Nonowning view to a sequence of contiguous elements.
+ *
+ * Spans encapsulate a pointer to a sequence of contiguous elements and its size
```
Query: What does "it" refer to in this sentence?

pan- replied:
_it_ refers to the sequence.

kjbracey replied:
Not sure that works - I get it from context, but don't think you can use "it" to reach into the "to" of a clause. It reads as either "encapsulates ((a pointer to a sequence of contiguous elements) and (the pointer's size))" or "encapsulates a pointer to ((a sequence of contiguous elements) and (the sequence's size))".

How about "Spans encapsulate the address and size of a sequence of contiguous elements"? That also avoids the indefiniteness of "a pointer" versus "its length".

Although I still like using the word pointer. How about "Spans encapsulate a pointer and length providing a view over a sequence of contiguous elements".

AnotherButler reviewed (approved):
I've made some copy edits. Please review them to make sure I didn't accidentally change the meaning of anything.

pan- commented on platform/Span.h:
```diff
@@ -49,29 +49,29 @@ class is_convertible
 
 /**
- * Non owning view to a sequence of contiguous elements.
+ * Nonowning view to a sequence of contiguous elements.
```
Shouldn't it be Non-owning ?

AnotherButler replied:
Use hyphens when the last letter of the suffix and the first letter of the prefix are the same [proto-oncology, non-negotiable, post-traumatic and so on]. There are some exceptions, such as when the suffix is a proper noun or an acronym. Then, use a hyphen regardless of the first letter of the suffix.

pan- replied:
That makes sense. Thanks Amanda. 
Looks like a lot of the CS literature spell it wrong.

pan- commented on platform/Span.h:
```diff
@@ -49,29 +49,29 @@ class is_convertible
  *
  * Span objects can be copied and assigned like regular value types with the help
- * of copy constructor and copy assignment (=) operator.
+ * of copy constructor and copy assignment (=) operators.
```
It is singular in this context; I can rephrase it: "with the help of the copy constructor or the copy assignment (=) operator "

pan- commented on platform/Span.h:
```diff
@@ -49,29 +49,29 @@ class is_convertible
+ * empty() informs whether there is any element in the sequence.
  *
- * Span can be sliced from the beginning of the sequence (first()), from the end
+ * You can splice Span from the beginning of the sequence (first()), from the end
```
_slice_ is more commonly used in CS in that context.

AnotherButler replied:
That was a typo on my part. My apologies.

pan- commented on platform/Span.h:
```diff
@@ -296,9 +296,9 @@ struct Span {
+     * @note For Span with a positive extent, this function is not accessible.
      *
-     * @note OtherElementType(*)[] should be convertible to ElementType(*)[].
+     * @note OtherElementType(*)[] is convertible to ElementType(*)[].
```
That's the enunciation of a requirement; not the enunciation of the fact. What form should be used ?

AnotherButler replied:
Oh, I see. What do you think of "must be"?

pan- replied:
That is more precise 👍 .

pan- committed:
Span: Fix documentation.

pan- commented:
@AnotherButler I've updated the documentation according to our discussion above.

AnotherButler reviewed (approved):
LGTM

cmonr commented:
@deepikabhavnani @c1728p9 @SenRamakri @kegilbert Could one of y'all take a look at this and OK it before CI?

cmonr commented:
Actually, gonna start CI with the hope that it's gets a final OK, since this this blocks #7822

/morph build

c1728p9 reviewed (approved):
This looks good to me. Thanks for bringing this main-line @pan-
